src/dynarray.h

```diff
@@ -0,0 +1,192 @@
+/****************************************************************************
+**
+**  This file is part of GAP, a system for computational discrete algebra.
+**
+**  Copyright of GAP belongs to its developers, whose names are too numerous
+**  to list here. Please refer to the COPYRIGHT file for details.
+**
+**  SPDX-License-Identifier: GPL-2.0-or-later
+**
+**  WARNING: This file should NOT be directly included. It is designed
+**  to instantiate a generic dynamic array type based on #defines that
+**  have to be set before including the file, and the file can be included
+**  repeatedly with different parameters.
+*/
+
+// Parameters:
+//
+// #define ELEM_TYPE type of elements
+// #define COMPARE comparison function for elements (optional)
+// #define ALLOC allocation function (optional)
+// #define DEALLOC deallocation function (optional)
+
+#ifdef __GNUC__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-function"
+#endif
+
+#define Array JOIN(ELEM_TYPE, Array)
+
+#define FN(sym) JOIN(Array, sym)
+
+#define JOIN(s1, s2) JOIN2(s1, s2)
+#define JOIN2(s1, s2) s1##s2
```

Comment:
These two macros also appear in `sortbase.h`. And in `debug.h`, we have them under a different name:
```
#define _intern_CONCAT_(X, Y)  X ## Y
#define _intern_CONCAT(X, Y)  _intern_CONCAT_(X,Y)
```
Perhaps we should just give in and move this to, say, `src/gaputils.h` and document it properly (this is not a change request for this PR, just a general thought. @ChrisJefferson ?)